[WIP] NNCLR Implementation

plumol opened the pull request:
## Description

<!--- What types of changes does your code introduce? -->
This implements the NNCLR loss function detailed in the paper, "[With a Little Help from My Friends: Nearest-Neighbor Contrastive Learning of Visual Representations](https://arxiv.org/abs/2104.14548)". It includes a new loss function NNCLRLoss and implements a support set queue as detailed in the reference paper.

Status: As of right now, it currently runs and compiles on Oscar, and I am running benchmarks for CIFAR10 and CIFAR100. Currently, SimCLR outperforms NNCLR on CIFAR10 (0.9241 vs. 0.8465), so I'm not 100% sure my implementation is accurate. I mainly used the same structure as the NX-Ent Loss.

<!--- Please link to an existing issue here if one exists. -->


## Checklist

<!-- - Go over all the following points, and put an `x` in all the boxes that apply. -->

- [x ] I have read the [**Contributing**](https://rbalestr-lab.github.io/stable-SSL.github.io/dev/contributing.html#) document.
- [x] The documentation is up-to-date with the changes I made (check build artifacts).
- [x] All tests passed, and additional code has been **covered with new tests**.
- [ ] I have added the PR to the [**RELEASES.rst**](../RELEASES.rst) file.

huguesva commented on stable_ssl/losses.py:
Could you maybe put the SupportSet class in utils ? Also would it make sense to use the NTXEntLoss but add the option to use a queue inside this class? Could be interesting to have the option to leverage this queue for all Losses defined on pairs also.

plumol replied:
Hi! I'm currently still running some benchmarks on it and clearing some config problems on my end- not resolved for now. 

That's a great point! I temporarily put it within the NNCLR class just to abstract it from everything else for now. I'll try to adapt it such that it follows a similar format to that of the memorybank module in the [lightly-ai repo](https://github.com/lightly-ai/lightly/blob/master/lightly/models/nnclr.py ).

huguesva reviewed (commented):
Hello @plumol ! thanks for the PR, did you manage to fix the performances on cifar? I have added some comments also

plumol commented:
I finally have a tested and cleaned version of NNCLR up and running. I've made some revisions and design choices:

- Moved the SupportQueue implementation into the utils class, but by the standard NNCLR implementation it defaults to outputting the nearest neighbor embeddings.
- Created a separate class for NNCLR, but it utilizes the NTXEntLoss in its forward() and performs queue calculations there. I figured it would be better to separate the two for now, but I can revisit it and merge them together if needed.

Next TODOs:

- Updating the documentation on the docs site
